Added support of textDocument/completion

alexsaveliev opened the pull request:
Addresses #55

felixfbecker commented on src/connection.ts:
```diff
@@ -241,6 +242,19 @@ export function registerLanguageHandler(connection: IConnection, strict: boolean
+			const result = await handler.getCompletions(params);
+			const exit = new Date().getTime();
+			console.error('completion items found', result && result.items ? result.items.length : 0, (exit - enter) / 1000.0);
+			return Promise.resolve(result);
```
If the function is `async`, just `return result`

alexsaveliev replied:
Done

felixfbecker commented on src/connection.ts:
```diff
@@ -241,6 +242,19 @@ export function registerLanguageHandler(connection: IConnection, strict: boolean
+			return Promise.resolve(result);
+		} catch (e) {
+			console.error(e);
+			return Promise.reject(e);
```
and `throw e`

alexsaveliev replied:
Done

felixfbecker commented on src/request-type.ts:
```diff
@@ -204,3 +204,11 @@ export namespace TextDocumentDidChangeNotification {
+
+export namespace TextDocumentCompletionRequest {
+	export const type: vscode.RequestType<vscode.TextDocumentPositionParams, vscode.CompletionList, void> = {
+		get method() {
```
is there a reason this can't just be a property?

alexsaveliev replied:
Style inspired by `vscode-languageserver` which holds original request definitions. For example,
https://github.com/Microsoft/vscode-languageserver-node/blob/5732a65f313e77af6ef55611e9a63dc4311a47c8/client/src/protocol.ts#L79

felixfbecker commented on src/test/test-utils.ts:
```diff
@@ -292,3 +292,15 @@ export function change(uri: string, text: string) {
 	});
 }
+
+export function completions(params: vscode.TextDocumentPositionParams, expected: string[], done: (err?: Error) => void) {
```
why is this function not promise-based/async instead of callbacks?

alexsaveliev replied:
I'd rather try to switch to async/promise-style as a part of separate tech debt task

felixfbecker commented on src/typescript-service.ts:
```diff
@@ -356,6 +362,38 @@ export class TypeScriptService implements LanguageHandler {
+			return null;
+		}
+
+		let items = [];
```
```ts
const items = completions.entries.map(entry => CompletionItem.create(completion.name))
```

alexsaveliev replied:
Done

felixfbecker commented on src/typescript-service.ts:
```diff
@@ -356,6 +362,38 @@ export class TypeScriptService implements LanguageHandler {
+
+		let items = [];
+		for (const completion of completions.entries) {
+			const item = CompletionItem.create(completion.name);
```
no other properties besides `name`, like `kind`?

alexsaveliev replied:
Added `kind`

felixfbecker commented on src/typescript-service.ts:
```diff
@@ -356,6 +363,39 @@ export class TypeScriptService implements LanguageHandler {
+			return null;
+		}
+
+		return CompletionList.create(completions.entries.map((item) => {
```
omit parens around `item`

felixfbecker commented on src/typescript-service.ts:
```diff
@@ -356,6 +363,39 @@ export class TypeScriptService implements LanguageHandler {
+			const kind = completionKinds.get(item.kind);
+			if (kind) {
+				ret.kind = kind;
+			}
```
Pretty sure you can just set the return value directly. Still a lot more properties though, documentation, type etc

felixfbecker commented on src/typescript-service.ts:
```diff
@@ -1358,3 +1398,22 @@ function pushall<T>(arr: T[], ...elems: (T | null | undefined)[]): number {
+completionKinds.set("text", CompletionItemKind.Text);
+completionKinds.set("text", CompletionItemKind.Unit);
+completionKinds.set("value", CompletionItemKind.Value);
+completionKinds.set("variable", CompletionItemKind.Variable);
```
Maps that don't change over time should just be an object hash, it's faster

felixfbecker commented on src/typescript-service.ts:
```diff
@@ -1358,3 +1398,22 @@ function pushall<T>(arr: T[], ...elems: (T | null | undefined)[]): number {
+completionKinds.set("reference", CompletionItemKind.Reference);
+completionKinds.set("snippet", CompletionItemKind.Snippet);
+completionKinds.set("text", CompletionItemKind.Text);
+completionKinds.set("text", CompletionItemKind.Unit);
```
duplicate

alexsaveliev commented:
@felixfbecker I have partially addressed your comments. For the rest I have tried to provide answers inline. I'd prefer to address code-style-related comments as a separate tech debt task if you don't mind

beyang commented on src/typescript-service.ts:
```diff
@@ -356,6 +363,42 @@ export class TypeScriptService implements LanguageHandler {
+		const column = params.position.character;
+		const fileName: string = util.uri2path(uri);
+
+		await this.projectManager.ensureFilesForReferences(uri);
```
Would `ensureFilesForHoverAndDefinition` be sufficient?

alexsaveliev replied:
You are right, done. Thanks!

beyang commented on src/test/language-server-test.ts:
```diff
@@ -1399,6 +1399,34 @@ var Node: {\n\
+		describe('completions', function () {
+			before(function (done: () => void) {
+				utils.setUp(newLanguageHandler(), {
+					'a.ts': `class A {
```
Can you add a test case for when the completion is defined in a separate file (i.e., imported)?

alexsaveliev replied:
Done

felixfbecker commented on src/typescript-service.ts:
```diff
@@ -1358,3 +1401,26 @@ function pushall<T>(arr: T[], ...elems: (T | null | undefined)[]): number {
+/**
+ * Maps string-based CompletionEntry::kind to enum-based CompletionItemKind
+ */
+const completionKinds: { [name: string]: CompletionItemKind } = {
```
If they are all exactly the same, you can do
```ts
ret.kind = CompletionItemKind[capitalize(item.kind)]
```

alexsaveliev replied:
We can't because `CompletionItemKind` is a **const** enum, see https://basarat.gitbooks.io/typescript/content/docs/enums.html#const-enums

felixfbecker replied:
Interesting, didn't know about this, thanks for the link!
Would be nice if the Microsoft used `preserveConstEnums` for the package.

beyang commented:
@felixfbecker anything else on this PR?

felixfbecker commented:
What about all the other properties on `CompletionItem`? Are they available, or should they be resolved though `completionItem/resolve`? Pretty sure the current implementation would be regression from the built-in VS Code TS IntelliSense

https://github.com/Microsoft/language-server-protocol/blob/master/versions/protocol-2-x.md#textDocument_completion

alexsaveliev commented:
> What about all the other properties on CompletionItem? Are they available, or should they be resolved though completionItem/resolve? Pretty sure the current implementation would be regression from the built-in VS Code TS IntelliSense

filling out the following fields: 
* label
* kind
* documentation
* detail
* sortText

>  ... should they be resolved though completionItem/resolve? 
Frankly, I had no plans to add support of `completionItem/resolve` now, unless requested

felixfbecker commented:
Okay, that looks much better! Approving. Did you check if TS provides a `textEdit` / `insertText`, `range`?

`resolve` would only make sense if computing the details was expensive (I don't think it is since the function is sync, so it probably has that cached and does not have to read a file for this)

prabirshrestha commented:
Could you release a new npm version so I can try this out.
